pin/pin.go

```diff
@@ -272,48 +303,58 @@ func (p *pinner) Flush() error {
 	p.lock.Lock()
 	defer p.lock.Unlock()
 
-	err := storeSet(p.dstore, directPinDatastoreKey, p.directPin.GetKeys())
-	if err != nil {
-		return err
-	}
+	ctx := context.TODO()
 
-	err = storeSet(p.dstore, recursePinDatastoreKey, p.recursePin.GetKeys())
-	if err != nil {
-		return err
+	internalPin := make(map[util.Key]struct{})
+	recordInternal := func(k util.Key) {
+		internalPin[k] = struct{}{}
 	}
 
-	err = storeIndirPin(p.dstore, indirectPinDatastoreKey, p.indirPin)
-	if err != nil {
-		return err
+	root := &mdag.Node{}
+	{
+		n, err := storeSet(ctx, p.dserv, p.directPin.GetKeys(), recordInternal)
+		if err != nil {
+			return err
+		}
+		if err := root.AddNodeLink(linkDirect, n); err != nil {
+			return err
+		}
 	}
-	return nil
-}
 
-// helpers to marshal / unmarshal a pin set
-func storeSet(d ds.Datastore, k ds.Key, val interface{}) error {
-	buf, err := json.Marshal(val)
-	if err != nil {
-		return err
+	{
+		n, err := storeSet(ctx, p.dserv, p.recursePin.GetKeys(), recordInternal)
+		if err != nil {
+			return err
+		}
+		if err := root.AddNodeLink(linkRecursive, n); err != nil {
+			return err
+		}
 	}
 
-	return d.Put(k, buf)
-}
+	{
+		n, err := storeMultiset(ctx, p.dserv, p.indirPin.GetRefs(), recordInternal)
+		if err != nil {
+			return err
+		}
+		if err := root.AddNodeLink(linkIndirect, n); err != nil {
+			return err
+		}
+	}
 
-func loadSet(d ds.Datastore, k ds.Key, val interface{}) error {
-	buf, err := d.Get(k)
+	k, err := p.dserv.Add(root)
 	if err != nil {
 		return err
 	}
-
-	bf, ok := buf.([]byte)
-	if !ok {
-		return errors.New("invalid pin set value in datastore")
+	internalPin[k] = struct{}{}
+	if err := p.dstore.Put(pinDatastoreKey, []byte(k)); err != nil {
+		return fmt.Errorf("cannot store pin state: %v", err)
```

Comment:
(related to https://github.com/ipfs/go-ipfs/pull/1225/files#r31384384 above) so yeah, here we'd put  the pins `root` into the repo `root` object, and store both, instead of just a k/v store put.